documentation/c05-packages.sil

```diff
@@ -1,33 +1,81 @@
 \begin{document}
 \chapter{SILE Packages}
 
-SILE comes with a number of packages which provide additional functionality.
-In fact, the actual “core” of SILE’s functionality is small and
-extensible, with most of the interesting features being provided by add-on
-packages. SILE packages are written in the Lua programming language, and can
-define new commands, change the way that the SILE system operates, or indeed
-do anything that is possible to do in Lua.
-
-As mentioned above, loading a package is done through the \code{\\script} command,
-which runs Lua code. By convention packages live in the \code{packages/} subdirectory
-of either your input file’s location, your current working directory, or SILE’s
-installation directory. For instance, we’ll soon be talking about the
-\code{grid} package, which normally can be found as
-\code{/usr/local/lib/sile/packages/grid/init.lua}. To load this, we’d say:
+SILE comes with a number of standard packages which provide additional functionality.
+In fact, the actual “core” of SILE’s functionality is small and extensible, with most of the interesting features being provided by add-on packages.
+SILE packages are written in the Lua programming language, and can define new commands, change the way that the SILE system operates, or indeed do anything that is possible to do in Lua.
+
+\section{Loading a package}
+
+Loading a package is done through the \code{\\use} command.
+By convention packages live in a \code{packages/}
+For instance, we’ll soon be talking about the \autodoc:package{grid} package, which normally can be found as \code{/usr/local/lib/sile/packages/grid/init.lua}.
+To load this, we’d say:
 
 \begin[type=autodoc:codeblock]{raw}
 \use[module=packages.grid]
 \end{raw}
 
-\autodoc:note{SILE searches for paths in a variety of
-directories: first, in the directory in which your input file is located,
-then the current working directory; next, if the environment variable
-\code{SILE_PATH} is set, it will look in that directory; then it will look in
-the standard installation directories \code{/usr/lib/sile} and
-\code{/usr/local/lib/sile}. Unlike TeX, it does not descend into subdirectories
-when looking for a file; if you have arranged your personal macro, class
-or package files into subdirectories, you will need to provide a full relative
-path to them.}
+\autodoc:note{
+By default SILE will look for packages in a variety of directories:
+
+\begin{enumerate}
+\item{The directory where your input source file is located.}
+\item{The current working directory.}
+\item{The environment variable \code{SILE_PATH}, if defined.}
+\item{The default Lua search path.}
+\item{The directories where SILE is installed on the system, \code{/usr/lib/sile} and
+\code{/usr/local/lib/sile}.}
+\end{enumerate}    
```

Comment:
As can be seen in the previous version, we were mentioning the search logic **3** times with different details...
The exact order is not clear to me and would have to be checked.

Reply:
According to the above command, there's some significant differences between my setups (Ubuntu PPA, Docker images, etc.). There's also `lua-libraries` and sometimes `lua_modules` subfolders in the mix, and the `sile` prefix too (e.g. even in the local path)...

Perhaps we should indeed be less detailed, e.g.

```
\item{The directory where your input source file is located.}
\item{The current working directory.}
\item{The environment variable \code{SILE_PATH}, if defined.}
\item{Various directories depending on where and how SILE is installed on your system.}
\item{The default Lua search path.}
```

N.B. The default Lua path _seems_ to come last.

Reply:
That working sounds fine.

And yes the default Lua (as defined by how the system Lua was compiled/configured) does come last. All the SILE specific bits whether project specific or related to vendored libraries or SILE itself are prepended to it.

And yes there is a huge a mount of variations here. At least 5 factors are involved each adding (or not) entirely different batches of paths. It took a long time and trial and error to work out something that worked robustly across platforms.

Reply:
What am I even talking about. The default Lua search path *needs* to be second to last as currently document. The SILE installation path should be last. Otherwise 3rd party packages installed via LuaRocks have no chance to *override* anything in the default SILE installation. Of course that runs the risk of users breaking stuff, but I think it gives more flexibility to the end user to know what is right for their project too.

If your installation doesn't reflect that then it's a bug.

Reply:
Indeed!
Done = simplified in a way that should be sufficient (removing explicit mention of a path in chap. 5 and 10).